snakemake_executor_plugin_interface/__init__.py

```diff
@@ -12,43 +17,105 @@ class CommonSettings:
-        ...
+
+        # Cut out early if we don't have custom parameters to add
+        if not self.has_executor_settings():
+            return
+
+        # Convenience handle
+        params = self._executor_settings_cls
+
+        # Assemble a new dataclass with the same fields, but with prefix
+        # fields are stored at dc.__dataclass_fields__
+        dc = copy.deepcopy(params)
+        for field in fields(params):
+            # Executor plugin dataclass members get prefixed with their
+            # name when passed into snakemake args.
+            prefixed_name = f"{self.prefix}_{field.name}"
+
+            # Since we use the helper function below, we
+            # need a new dataclass that has these prefixes
+            del dc.__dataclass_fields__[field.name]
+            field.name = prefixed_name
+            dc.__dataclass_fields__[field.name] = field
+
+        # When we get here, we have a namespaced dataclass.
+        # If there is overlap in snakemake args, it should error
+        _add_dataclass_options(dc, argparser)
+
+    def has_executor_settings(self):
+        """Determine if a plugin defines custom executor settings"""
+        return self._executor_settings_cls is not None and not isinstance(
+            self._executor_settings_cls, ExecutorSettingsBase
+        )
 
     def get_executor_settings(self, args) -> Optional[ExecutorSettingsBase]:
-        """Return an instance of self.executor_settings with values from args."""
-        ...
+        """Return an instance of self.executor_settings with values from args.
+
+        This helper function will select executor plugin namespaces arguments
+        for a dataclass. It allows us to pass them from the custom executor ->
+        custom argument parser -> back into dataclass -> snakemake.
+        """
+        if not self.has_executor_settings():
+            return ExecutorSettingsBase()
 
+        # We will parse the args from snakemake back into the dataclass
+        dc = self._executor_settings_cls
+
+        # Iterate through the args, and parse those in the namespace
+        kwargs = {}
+
+        # These fields will have the executor prefix
+        for field in fields(dc):
+            # This is the actual field name without the prefix
+            name = field.name.replace(f"{self.name}_", "", 1)
+            value = getattr(args, field.name, None)
+
+            # This will only add instantiated values, and
+            # skip over dataclasses._MISSING_TYPE and similar
+            if isinstance(value, ArgTypes):
+                kwargs[name] = value
+
+        # At this point we want to convert back to the original dataclass
+        return dc(**kwargs)
 
 
 class ExecutorPluginRegistry:
     """This class is a singleton that holds all registered executor plugins."""
-    _instance = None
-
-    def __new__(cls):
-        if cls._instance is None:
-            cls._instance = super().__new__(cls)
-        return cls._instance
```

Comment:
Why deleting this? It is the mechanism to obtain the singleton instance. Or am I missing something?

Reply:
It just didn’t work to create the class - I don’t remember the exact error (falling asleep now but can report tomorrow ).